<a href="https://www.kaggle.com/code/beatafaron/fmcg-2022-24-eda-key-business-insights?scriptVersionId=267719342" target="_blank"><img align="left" alt="Kaggle" title="Open in Kaggle" src="https://kaggle.com/static/images/open-in-kaggle.svg"></a>

# 1. Introduction

This notebook presents a comprehensive exploratory analysis of a selected SKU (`MI-006`, Milk category) using synthetic FMCG sales data from **2022 to 2024**.

It focuses on uncovering patterns in sales behavior, promotions, stockouts, and seasonality – all within the lifecycle of a single SKU.  
All logic is implemented with **modular and reusable functions**, allowing easy extension to other SKUs, brands, or regions.

---

### What You’ll Find Here

-  Calendar-based feature engineering  
-  Stock availability and stockout detection  
-  Promotion effectiveness across channels and lifecycle phases  
-  Sales trends, momentum, and seasonal decomposition  
-  Business insights structured by **product lifecycle stage**

> ⚠️ _Note: This is a synthetic dataset. In real-life FMCG environments, delivery delays, promotions, and stockouts often show stronger and more correlated effects._


##  Table of Contents

1. [Introduction](#1-introduction)  
2. [Data Overview](#2-data-overview)  
3. [Calendar Features](#3-calendar-features)  
4. [Daily Volume Check](#4-daily-volume-check)  
5. [Weekly Demand Decomposition](#5-weekly-demand-decomposition)  
6. [Promotions & Holidays](#6-promotions--holidays)  
7. [Stockouts Analysis](#7-stockouts-analysis)  
8. [SKU Lifecycle](#8-sku-lifecycle)  
9. [Weekly Sales Momentum](#9-weekly-sales-momentum)  
10. [Lead Time vs Demand](#10-lead-time-vs-demand)  
11. [Category Seasonality](#11-category-seasonality)  
12. [Feature Engineering](#12-feature-engineering)  
13. [Feature Distributions](#13-feature-distributions)  
14. [Summary & Recommendations](#14-summary--recommendations)


# 2. Data Overview

We begin by loading the **daily-level transactional dataset**.  
The data contains synthetic FMCG sales transactions for multiple SKUs across different **channels** and **regions** between **2022 and 2024**.

For this analysis, we focus on **SKU `MI-006` (Milk category)**.  
All functions are parameterized to allow easy reuse for other SKUs, brands, or regions.


In [1]:
# === Standard libraries ===
import warnings
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns


# === Custom utility modules (EDA plots) ===
from utils import (
    plot_gradient_bar,
    plot_gradient_bar_grouped,
    save_and_display,
    plot_sku_lifecycle_timeline,
    plot_transactions_per_day,
    plot_daily_transactions_for_sku,
    plot_weekly_sales_overlay_plotly,
    plot_grouped_bar,
    plot_event_timeline_plotly,
    plot_channel_event_probabilities,
    plot_line_generic,
    plot_lifecycle_bar,
    plot_dual_axis_line,
    plot_box_numeric_vs_units,
    decompose_weekly_sales_plotly,
    plot_monthly_seasonality,
    plot_feature_distributions_with_trend
)

# === Configuration ===
warnings.filterwarnings('ignore')
pd.options.display.max_columns = None

In [2]:
df = pd.read_csv("/kaggle/input/fmcg-daily-sales-data-to-2022-2024/FMCG_2022_2024.csv")
df["date"] = pd.to_datetime(df["date"])
df.head()

,date,sku,brand,segment,category,channel,region,pack_type,price_unit,promotion_flag,delivery_days,stock_available,delivered_qty,units_sold
0,2022-01-21,MI-006,MiBrand1,Milk-Seg3,Milk,Retail,PL-Central,Multipack,2.38,0,1,141,128,9
1,2022-01-21,MI-006,MiBrand1,Milk-Seg3,Milk,Retail,PL-North,Single,1.55,1,3,0,129,0
2,2022-01-21,MI-006,MiBrand1,Milk-Seg3,Milk,Retail,PL-South,Carton,4.00,0,5,118,161,8
3,2022-01-21,MI-006,MiBrand1,Milk-Seg3,Milk,Discount,PL-Central,Single,5.16,0,2,81,114,7
4,2022-01-21,MI-006,MiBrand1,Milk-Seg3,Milk,Discount,PL-North,Single,7.66,0,4,148,204,12


In [3]:
# Basic dataset shape
print(f"Dataset shape: {df.shape[0]:,} rows × {df.shape[1]:,} columns")

# Quick info on columns
df.info()

# Time range
print(f"Date range: {df['date'].min().date()} → {df['date'].max().date()}")

Dataset shape: 190,757 rows × 14 columns
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 190757 entries, 0 to 190756
Data columns (total 14 columns):
 #   Column           Non-Null Count   Dtype         
---  ------           --------------   -----         
 0   date             190757 non-null  datetime64[ns]
 1   sku              190757 non-null  object        
 2   brand            190757 non-null  object        
 3   segment          190757 non-null  object        
 4   category         190757 non-null  object        
 5   channel          190757 non-null  object        
 6   region           190757 non-null  object        
 7   pack_type        190757 non-null  object        
 8   price_unit       190757 non-null  float64       
 9   promotion_flag   190757 non-null  int64         
 10  delivery_days    190757 non-null  int64         
 11  stock_available  190757 non-null  int64         
 12  delivered_qty    190757 non-null  int64         
 13  units_sold       190757 non-null 

###  Data sanity check

In [4]:
# checking the continuity of dates
date_range = pd.date_range(start=df['date'].min(), end=df['date'].max(), freq='D')
missing_dates = date_range.difference(df['date'].dropna().unique())

print("Number of missing dates:", len(missing_dates))
print("Missing dates:", missing_dates)

Number of missing dates: 0
Missing dates: DatetimeIndex([], dtype='datetime64[ns]', freq='D')


In [5]:
# Get all unique dates and SKUs
all_dates = pd.date_range(start=df['date'].min(), end=df['date'].max(), freq='D')
all_skus = df['sku'].unique()

# Create a full date × SKU grid
full_grid = pd.DataFrame([(date, sku) for date in all_dates for sku in all_skus], columns=['date', 'sku'])

# Get existing (date, SKU) combinations
existing = df[['date', 'sku']].drop_duplicates()

# Merge and find missing combinations
merged = full_grid.merge(existing, on=['date', 'sku'], how='left', indicator=True)
missing = merged[merged['_merge'] == 'left_only'].drop(columns=['_merge'])

# Result
print(f"Missing (date, SKU) combinations: {len(missing)}")
display(missing.head(5))


Missing (date, SKU) combinations: 7335


,date,sku
1,2022-01-21,MI-026
2,2022-01-21,YO-029
3,2022-01-21,YO-005
4,2022-01-21,YO-012
5,2022-01-21,YO-014


In [6]:
import plotly.graph_objects as go

# Ensure 'date' is datetime
df['date'] = pd.to_datetime(df['date'])

# Prepare data
skus = sorted(df['sku'].unique())

# Create figure
fig = go.Figure()

for i, sku in enumerate(skus):
    dates_for_sku = (
        df[df['sku'] == sku]['date']
        .sort_values()
        .unique()
    )
    fig.add_trace(go.Scatter(
        x=dates_for_sku,
        y=[sku] * len(dates_for_sku),
        mode='lines',
        line=dict(width=4),
        name=sku,
        hovertemplate=f"SKU: {sku}<br>Date: %{{x|%Y-%m-%d}}<extra></extra>",
        showlegend=False
    ))

# Calculate dynamic height
dynamic_height = 100 + len(skus) * 20  # adjust multiplier for spacing

# Layout formatting
fig.update_layout(
    title="Product Lifecycle Timeline — One Line per SKU",
    xaxis_title="Date",
    yaxis=dict(
        categoryorder="array",
        categoryarray=skus,
        side="right"
    ),
    plot_bgcolor='white',
    xaxis=dict(showgrid=True, gridcolor='lightgrey'),
    height=dynamic_height
)

# Display full height (update iframe height as well)
save_and_display(fig, "sku_lifecycle_timeline_full", iframe_height=dynamic_height + 100)


'images/sku_lifecycle_timeline_full.html'

>The timeline visualization reveals clear patterns in SKU availability over time. Some products have continuous presence throughout the dataset, while others exhibit shorter or intermittent lifecycles. These gaps often reflect the natural introduction or discontinuation of products rather than missing data.
>This observation is crucial for accurate modeling. It suggests that imputation strategies (e.g., filling missing sales with zeros) should be applied only within each product’s active period to avoid introducing artificial demand for products that were not available in the market.
>
>By aligning data preprocessing with product lifecycle logic, we ensure more realistic forecasts and better business decisions.

In [7]:
import plotly.graph_objects as go

# Prepare data
daily_counts = df.groupby("date").size().reset_index(name="row_count")

# Create figure
fig = go.Figure()
fig.add_trace(go.Scatter(
    x=daily_counts["date"],
    y=daily_counts["row_count"],
    mode="lines",
    line=dict(width=2, color="#4d7cfe"),
    hovertemplate="Date: %{x|%Y-%m-%d}<br>Rows: %{y:,}<extra></extra>"
))

# Layout
fig.update_layout(
    title="Number of Transactions per Day",
    xaxis_title="Date",
    yaxis_title="Row Count",
    plot_bgcolor='white',
    xaxis=dict(showgrid=True, gridcolor='lightgrey'),
    yaxis=dict(showgrid=True, gridcolor='lightgrey'),
    height=400
)

save_and_display(fig, "transactions_per_day")


'images/transactions_per_day.html'

# 3. Calendar Features

To capture seasonal and temporal patterns in demand, we enrich the dataset with calendar-based features using a reusable function.

The function `add_calendar_features()` adds the following features at the daily level:
- **Day of week** (0 = Monday, ..., 6 = Sunday)
- **Weekend flag**
- **Month, quarter, year**
- **Start/end of month**
- **Start/end of quarter**
- **Week start date** (`week`) and ISO **week number**
- **Meteorological season** (Winter, Spring, Summer, Autumn)

Additionally, we manually flag selected **holiday peak shopping days**:
- Christmas (Dec 22–24)
- Easter (Apr 5–7)
- All Saints (Oct 30–31)
- May Holiday (Apr 30–May 2)
- New Year's Eve (Dec 30–31)

These flags help capture **short-term demand spikes** often seen in FMCG.


In [8]:
def add_calendar_features(df: pd.DataFrame, date_col: str = "date") -> pd.DataFrame:
    """
    Adds standard calendar-based features to a DataFrame.
    
    Parameters:
        df (pd.DataFrame): Input DataFrame with a datetime column.
        date_col (str): Name of the datetime column (default: 'date').

    Returns:
        pd.DataFrame: DataFrame with enriched calendar features.
    """
    df = df.copy()
    df[date_col] = pd.to_datetime(df[date_col])

    # Basic calendar breakdown
    df["day_of_week"] = df[date_col].dt.dayofweek
    df["is_weekend"] = df["day_of_week"].isin([5, 6])
    df["month"] = df[date_col].dt.month
    df["year"] = df[date_col].dt.year
    df["quarter"] = df[date_col].dt.quarter
    df["is_start_of_month"] = df[date_col].dt.is_month_start
    df["is_end_of_month"] = df[date_col].dt.is_month_end
    df["is_quarter_start"] = df[date_col].dt.is_quarter_start
    df["is_quarter_end"] = df[date_col].dt.is_quarter_end

    # Week start (Monday) and ISO week number
    df["week"] = df[date_col] - pd.to_timedelta(df[date_col].dt.weekday, unit='D')
    df["week_number"] = df["week"].dt.isocalendar().week

    # Map month to meteorological seasons
    month_to_season = {
        12: "winter", 1: "winter", 2: "winter",
        3: "spring", 4: "spring", 5: "spring",
        6: "summer", 7: "summer", 8: "summer",
        9: "autumn", 10: "autumn", 11: "autumn"
    }
    df["season"] = df["month"].map(month_to_season)
    return df

# Apply function
df = add_calendar_features(df, date_col="date")

# Define manual holiday peak days (fixed month-day combinations)
holiday_dates = {
    "Christmas": ["12-22", "12-23", "12-24"],
    "Easter": ["04-05", "04-06", "04-07"],
    "All Saints": ["10-30", "10-31"],
    "May Weekend": ["04-30", "05-01", "05-02"],
    "New Year's Eve": ["12-30", "12-31"]
}

# Flatten all holiday date strings
flat_dates = sum(holiday_dates.values(), [])

# Expand to full date strings for all years in the dataset
holiday_peak = [f"{year}-{d}" for year in df["year"].unique() for d in flat_dates]

# Flag holiday peaks
df["date_str"] = df["date"].dt.strftime("%Y-%m-%d")
df["is_holiday_peak"] = df["date_str"].isin(holiday_peak)
df.drop("date_str", axis=1, inplace=True)


### Weekly Aggregation 

Most time-series analyses in this notebook are conducted on **weekly granularity**, rather than daily.

We aggregate the dataset by:
- `week` (starting Monday)
- `sku`

This allows us to analyze:
- Long-term trends
- Promotion and lifecycle patterns
- Seasonality and momentum


In [9]:
# Create weekly-level dataset: aggregate by SKU, week, channel, and region
weekly_df = (
    df.groupby(["sku", "week", "channel", "region"])
    .agg({
        "units_sold": "sum",
        "stock_available": "mean",
        "promotion_flag": "max",
        "price_unit": "mean",
        "delivery_days": "mean",
        "is_holiday_peak": "max"
    })
    .reset_index()
)

In [10]:
# Week (1–52)
weekly_df["week_number"] = weekly_df["week"].dt.isocalendar().week

# Month and year
weekly_df["month"] = weekly_df["week"].dt.month
weekly_df["year"] = weekly_df["week"].dt.year

# Holiday Flag – checks if the week contains a holiday (from original df)
holiday_weeks = df[df["is_holiday_peak"] == 1]["week"].unique()
weekly_df["is_holiday_week"] = weekly_df["week"].isin(holiday_weeks).astype(int)

# Simple Seasonal Flags
weekly_df["is_summer"] = weekly_df["month"].isin([6, 7, 8]).astype(int)
weekly_df["is_winter"] = weekly_df["month"].isin([12, 1, 2]).astype(int)

### Daily  Volume  Check


### Daily Transaction Volume – SKU Presence Check

We plot the daily number of transactions for SKU `MI-006` to verify:

- Data density and continuity
- Any gaps or low-activity periods
- Seasonality in data collection

A 7-day moving average helps smooth out weekly fluctuations.


> To better observe the overall trend and reduce daily noise, we applied a 7-day moving average. This helps highlight longer-term changes in transaction volume and makes seasonal patterns easier to detect.


In [11]:
plot_daily_transactions_for_sku(df, sku="MI-006")


'images/daily_transactions_MI-006.html'

#  4. Decomposition of Weekly Demand

###  SKU: MI-006

To better understand the sales behavior of SKU `MI-006`, we use a **multiplicative seasonal decomposition** on its weekly sales time series.

The decomposition isolates:
- **Trend**: Long-term direction (e.g. product maturity or decline)
- **Seasonality**: Weekly demand cycles
- **Residual**: Irregular fluctuations (e.g. stockouts, anomalies)

We use a reusable function `decompose_weekly_sales()` to perform this breakdown.  
This makes the logic easy to apply across other SKUs in future use cases.


In [12]:
decompose_weekly_sales_plotly(weekly_df, sku="MI-006")


'images/weekly_decomposition_MI-006.html'


- **Trend**: Demand rose significantly in early 2022, peaking around mid-year, followed by a gradual decline and stabilization. This likely reflects either product launch effects or seasonal market saturation.

- **Seasonality**: A clear **weekly seasonal pattern** is visible, repeating consistently across the entire period. This may correspond to retail cycles (e.g. weekend shopping behavior, supply restocking rhythms).

- **Residuals**: While the residual component remains mostly stable, there are some outliers — likely caused by **promotional spikes, holidays, or unusual external events**.


# 5. Promotion & Holidays 

### Promotion vs Holidays – Sales Trend & Interaction (SKU: MI-006)

We explore how **promotions** and **holiday periods** impact sales of SKU `MI-006` on two levels:

1. **Weekly Sales Trend**  
   We visualize weekly units sold, highlighting promo and holiday weeks with overlays.  
   This helps reveal demand surges around commercial peaks.

2. **Promo × Holiday Interaction**  
   Using a grouped bar chart, we compare average weekly sales across four segments:
   - Promo during holiday peak
   - Promo outside holiday
   - Non-promo during holiday
   - Non-promo, non-holiday (baseline)

This breakdown helps separate **promo uplift** from **holiday-driven spikes**.


In [13]:
plot_weekly_sales_overlay_plotly(weekly_df, sku="MI-006")


'images/weekly_sales_overlay_MI-006.html'

In [14]:

# Filter SKU
df_sku = weekly_df[weekly_df["sku"] == "MI-006"].copy()

# Pivot table: is_holiday_peak vs promotion_flag
pivot = (
    df_sku.groupby(["is_holiday_peak", "promotion_flag"])["units_sold"]
    .mean()
    .unstack(fill_value=0)
)
plot_grouped_bar(
    labels=pivot.index.tolist(),   # <- dynamicznie z pivotu
    series_dict={
        "No Promotion": pivot[0].tolist(),
        "Promotion": pivot[1].tolist()
    },
    title="Effect of Promotion and Holiday on Average Sales"
)


'images/grouped_bar_chart.html'


>* Products on **promotion** consistently show **higher average sales** than those not on promotion.
>* During **holiday periods**, the combination of a promotion and a peak season leads to the **highest sales uplift** (22.82 units on average).
>* The base level (no promotion, no holiday) shows the lowest average sales (14.42 units).
>* Interestingly, **promotion outside holidays** performs even better than holiday-only sales, suggesting that **promotion is the stronger driver**.

>**Conclusion:**
For SKU `MI-006`, promotions significantly increase sales, especially when aligned with holiday periods. However, promotions also perform strongly even outside seasonal peaks — making them a valuable sales lever across the calendar.

# 6. Stockouts Analysis


###  Stockouts – Detection, Streaks & Sales Impact (SKU: MI-006)

Product unavailability (stockouts) can lead to lost sales and reduced customer loyalty.

In this section, we analyze stockouts for SKU `MI-006` through:
-  **Daily stockout timeline** – raw events and 7-day moving average
-  **Stockout × Promotion interaction** – opportunity loss
-  **Stockout streak detection** – consecutive out-of-stock periods
-  **Sales impact** – comparing sales on stockout vs available days

This helps evaluate both **operational risks** and **sales leakage** from poor availability.


### Daily Stockout Events (Smoothed Timeline)

In [15]:
df_sku = df[df["sku"] == "MI-006"].copy()
df_sku["is_stockout"] = (df_sku["stock_available"] == 0).astype(int)

daily_counts = df_sku.groupby("date")["is_stockout"].sum()
smoothed = daily_counts.rolling(window=7).mean()


plot_event_timeline_plotly(
    dates=daily_counts.index,
    daily_values=daily_counts.values,
    smoothed_values=smoothed.values,
    event_label="Stockouts",
    sku="MI-006"
)


'images/stockouts_timeline_MI-006.html'

### Stockout Streaks – Consecutive Gaps

###  Stockout Streaks – Consecutive Gaps (≥3 Days)

Beyond isolated events, it's important to detect **stockout streaks** — consecutive days with zero availability.

Using a reusable function `find_stockout_streaks()`, we identify multi-day gaps in availability for SKU `MI-006`.

This helps evaluate operational risk and estimate **unmet demand potential** during long stockouts.


In [16]:
from typing import List, Tuple, Optional

def find_stockout_streaks(
    df: pd.DataFrame,
    sku: str,
    channel: Optional[str] = None,
    min_days: int = 3
) -> List[Tuple[pd.Timestamp, pd.Timestamp, int]]:
    """
    Identifies periods of consecutive stockout days (stock_available == 0)
    for a specific SKU (and optional channel).

    Parameters:
        df (pd.DataFrame): Daily-level DataFrame with columns ['date', 'sku', 'channel', 'stock_available'].
        sku (str): The SKU to analyze.
        channel (str, optional): If provided, filter by this channel as well.
        min_days (int): Minimum number of consecutive stockout days to report.

    Returns:
        List of tuples: [(start_date, end_date, duration_days), ...]
    """
    temp_df = df[df['sku'] == sku].copy()
    if channel:
        temp_df = temp_df[temp_df['channel'] == channel]

    temp_df['date'] = pd.to_datetime(temp_df['date'])

    if temp_df.empty:
        return []

    daily_stock = (
        temp_df.groupby('date')['stock_available']
        .min()
        .reset_index()
    )

    stockout_days = daily_stock[daily_stock['stock_available'] == 0]['date'].sort_values()

    if stockout_days.empty:
        return []

    streaks = []
    start = stockout_days.iloc[0]
    prev = start

    for current in stockout_days[1:]:
        if (current - prev).days == 1:
            prev = current
        else:
            duration = (prev - start).days + 1
            if duration >= min_days:
                streaks.append((start, prev, duration))
            start = current
            prev = current

    duration = (prev - start).days + 1
    if duration >= min_days:
        streaks.append((start, prev, duration))

    return streaks

# Example run for Retail channel
streaks = find_stockout_streaks(df, sku="MI-006", channel="Retail", min_days=3)
for s in streaks:
    print(f"• From {s[0].date()} to {s[1].date()} → {s[2]} days")


• From 2022-05-15 to 2022-05-18 → 4 days


### Stockouts by Channel (with Holiday Overlay)

This visualization shows how stockouts varied by **sales channel** (e.g. Retail, Discount, etc.) for SKU `MI-006`.

We calculate the **daily stockout probability per channel**, using:
- 7-day rolling average for smoothing
- Holiday shopping periods overlaid as vertical lines

This helps identify operational gaps or demand surges that affect availability in specific distribution networks.


In [17]:
sku = "MI-006"
df_sku = df[df['sku'] == sku].copy()

# Daily stockout flag per channel
grouped = (
    df_sku.groupby(['date', 'channel'])['stock_available']
    .min()
    .reset_index()
)
grouped['is_stockout'] = grouped['stock_available'].eq(0).astype(int)

# Pivot: channels as columns
stockouts_pivot = grouped.pivot(index='date', columns='channel', values='is_stockout').fillna(0)
smoothed = stockouts_pivot.rolling(window=7).mean()

holiday_dates = df_sku.loc[df_sku['is_holiday_peak'] == 1, 'date'].unique()

# Wywołanie
plot_channel_event_probabilities(
    dates=smoothed.index,
    channel_df=smoothed,
    holiday_dates=holiday_dates,
    title=f"Stockout Probability by Channel — SKU: {sku}"
)


'images/channel_event_probabilities.html'

>The chart shows weekly stockout probability (7-day average) for each channel:
>
>- **Discount and E-commerce channels** exhibit sharper spikes and more frequent disruptions, especially near national holidays.
>- **Retail channel** remains relatively stable, suggesting better inventory coverage or supply chain prioritization.
>- **Holiday peaks** often align with increased stockout risk, emphasizing the need for improved forecasting and replenishment during these critical periods.

>  These insights can help inform **channel-specific safety stock policies**, proactive logistics planning, and demand-driven promotions.


# 7.  Lifecycle Analysis

### SKU Age, Stage & Promo Effectiveness (MI-006)

Every product has a natural lifecycle: **Growth → Maturity → Decline**.  
In this section, we assess the lifecycle of SKU `MI-006` and evaluate:

1. **Sales curve vs product age**  
   → How average weekly sales change as the SKU matures.

2. **Sales vs lifecycle stage**  
   → How demand differs in the Growth, Mature, and Decline phases.

3. **Promotion impact by lifecycle stage**  
   → Are promotions equally effective at all stages?

---

###  Step 1: Calculate SKU Age & Lifecycle Stage

Using `sku_age_weeks`, we define lifecycle phases:

| Age (weeks) | Stage     |
|-------------|-----------|
| 0–12        | Growth    |
| 13–52       | Mature    |
| 53+         | Decline   |

The classification is added via a custom function `classify_lifecycle()`.

---

###  Step 2: Plot Sales Curve vs SKU Age

Function used: `plot_sku_lifecycle()`  
This shows **how demand evolves** as the SKU ages.

---

###  Step 3: Sales by Lifecycle Stage

Function used: `plot_lifecycle_sales_for_sku()`  
Visualizes **average units sold** in each lifecycle stage.

---

###  Step 4: Promo Effectiveness by Lifecycle

Function used: `plot_promo_effect_by_lifecycle()`  
Analyzes **how promotions perform across lifecycle stages**.  
Highlights whether uplift is strongest during Growth, or fades in Decline.

---

>  Business Insight: Promotions tend to deliver the most impact in the **Mature** stage.  
> The **Growth** phase shows natural volume growth even without heavy promo activity.


### Compute SKU Age & Lifecycle Stage

In [18]:
# Get first week of each SKU
sku_first_week = weekly_df.groupby("sku")["week"].min().to_dict()

# Add SKU age in weeks
weekly_df["sku_age"] = weekly_df.apply(
    lambda row: (row["week"] - sku_first_week[row["sku"]]).days // 7,
    axis=1
)

# Lifecycle phase classification
def classify_lifecycle(age_weeks):
    if age_weeks <= 12:
        return "Growth"
    elif age_weeks <= 52:
        return "Mature"
    else:
        return "Decline"

weekly_df["lifecycle_stage"] = weekly_df["sku_age"].apply(classify_lifecycle)


### Sales by Lifecycle Phase 

In [19]:
sku = "MI-006"

# Filter for selected SKU
df_sku = weekly_df[weekly_df["sku"] == sku].copy()

# Aggregate avg units sold by sku_age
sku_lifecycle = (
    df_sku.groupby("sku_age")["units_sold"]
    .mean()
    .reset_index()
    .rename(columns={"units_sold": "avg_units_sold"})
)

In [20]:
plot_line_generic(
    x=sku_lifecycle['sku_age'],
    y=sku_lifecycle['avg_units_sold'],
    title=f"SKU Lifecycle – Avg Weekly Sales vs SKU Age – {sku}",
    xaxis_title="SKU Age (weeks)",
    yaxis_title="Avg Units Sold",
    height=580,        
    y_dtick=10         
)


'images/line_chart_generic.html'


>- **Rapid growth phase** in the first ~15 weeks, peaking around week 10–12. This is typical for new product launches supported by initial distribution push or marketing.
>- A **steady decline and stabilization** follows, with demand settling into a baseline between weeks 20 and 100.
>-  Slight late-stage fluctuation, possibly driven by seasonality, promotions, or slow phase-out activities.

> 💡 Lifecycle-aware features (e.g., SKU age, launch phase flags) are valuable additions to demand forecasting models — they help contextualize performance over time and prevent overfitting to transient peaks.


### Sales by Lifecycle Stage

In [21]:
# Agregacja w notebooku
df_sku = weekly_df[weekly_df['sku'] == "MI-006"].copy()

lifecycle_summary = (
    df_sku.groupby("lifecycle_stage")["units_sold"]
    .mean()
    .reset_index()
    .sort_values("lifecycle_stage")
)

labels = lifecycle_summary["lifecycle_stage"].tolist()
values = lifecycle_summary["units_sold"].tolist()

# Wykres
plot_lifecycle_bar(labels, values, sku="MI-006")


'images/lifecycle_bar_MI-006.html'

>- 🔵 **Mature**: This phase showed the **highest average sales**, indicating it’s the product’s most stable and profitable period.
>- 🟠 **Growth**: Demand increased steadily as the product gained traction in the market — strong overall performance.
>- 🟢 **Decline**: Although lower in volume, sales remained **moderate and consistent**, which can still be valuable depending on margin or stock levels.

> These insights help tailor promotion timing, forecast assumptions, and lifecycle-aware replenishment strategies.


### Promo Effectiveness by Lifecycle

In [22]:
# Przygotowanie danych pod lifecycle stage
df_sku = weekly_df[weekly_df["sku"] == "MI-006"].copy()

# Upewniamy się, że kolumna istnieje
if "promotion_flag" not in df_sku.columns:
    raise KeyError("'promotion_flag' column not found in dataset.")

pivot = (
    df_sku.groupby(["lifecycle_stage", "promotion_flag"])["units_sold"]
    .mean()
    .unstack(fill_value=0)
)

# Wykres przy użyciu plot_grouped_bar
plot_grouped_bar(
    labels=pivot.index.tolist(),
    series_dict={
        "No Promotion": pivot[0].tolist(),
        "Promotion": pivot[1].tolist()
    },
    title="Promotion Effectiveness by Lifecycle Stage — SKU: MI-006",
    yaxis_title="Avg Weekly Units Sold"
)


'images/grouped_bar_chart.html'

>-  **Mature stage** showed the **strongest uplift** from promotions — confirming it's the most valuable window for promo campaigns.
>-  **Growth stage** also benefited significantly, likely due to higher responsiveness during market expansion.
>-  Even in the **Decline stage**, promotions still triggered noticeable increases in demand, although from a lower baseline.

>  This analysis supports the idea that **timing matters** — and that mature SKUs benefit the most from well-targeted promotions.

#  8. Trend & Momentum

### Weekly Sales Trend & Momentum (SKU: MI-006)

The chart below visualizes:
- The smoothed **weekly sales trend** using a 3-week moving average
- The **week-over-week % momentum** using a secondary axis

Momentum can help detect early inflection points before volume shifts become visible.


In [23]:
sku = "MI-006"

# Filter for selected SKU
df_sku = weekly_df[weekly_df["sku"] == sku].copy().sort_values("week")

# Calculate rolling mean (3-week moving average)
df_sku['rolling_mean_3'] = df_sku['units_sold'].rolling(window=3).mean()

# Calculate momentum (week-over-week % change)
df_sku['momentum'] = df_sku['units_sold'].pct_change() * 100

plot_dual_axis_line(
    x=df_sku['week'],
    y1=df_sku['rolling_mean_3'],
    y2=df_sku['momentum'],
    y1_label="Sales Trend (3-week MA)",
    y2_label="Momentum (WoW %)",
    title=f"Weekly Sales Trend and Momentum — SKU: {sku}",
    y1_color="steelblue",
    y2_color="darkorange"
)


'images/dual_axis_line.html'


> To uncover product-level dynamics, we analyzed **weekly average sales** alongside their **momentum** (week-over-week percentage change) for selected SKUs.
>* Each SKU exhibits distinct sales behavior over time, reflecting differences in lifecycle stages, seasonality, and demand patterns.
>* **Sales trend (blue line)** shows the smoothed level of weekly sales volume.
>* **Momentum (orange dashed line)** captures the **rate of change** in weekly sales:

>  * Positive momentum ➜ growing demand
>  * Negative momentum ➜ declining or stabilizing demand


>* Some SKUs show clear **growth phases**, with sustained positive momentum followed by stabilization or decline.
>* **Sharp negative momentum spikes** often precede periods of stagnation or indicate possible stockouts or promotional burnout.
>* Monitoring momentum helps detect **early inflection points**, allowing for proactive decisions in production, inventory, or marketing.

>This combined view of **trend + momentum** provides a richer understanding than volume alone — helping stakeholders react not just to what happened, but to where the product is likely headed.

#  9. Lead Time vs Demand


###  Operational Insight – Lead Time vs Demand (SKU: MI-006)

To explore the potential **supply chain impact** on demand,  
we analyze the relationship between **delivery lead time** and **units sold**.

Longer delivery times can reduce customer satisfaction and lead to:
- Lost sales
- Substitution with other products
- Retailer penalties (e.g. in FMCG or e-commerce)

We use a boxplot and correlation coefficient to check for patterns.


In [24]:
sku = "MI-006"
df_sku = df[df['sku'] == sku].copy()
df_sku['delivery_days'] = df_sku['delivery_days'].astype(int)

# Correlation
correlation = df_sku[['delivery_days', 'units_sold']].corr().iloc[0, 1]
print(f"Correlation between delivery_days and units_sold: {correlation:.2f}")

# Plot
plot_box_numeric_vs_units(
    x_values=df_sku['delivery_days'],
    y_values=df_sku['units_sold'],
    xaxis_title="Delivery Days",
    yaxis_title="Units Sold",
    title=f"Lead Time vs Units Sold — SKU: {sku}"
)


Correlation between delivery_days and units_sold: -0.01


'images/box_numeric_vs_units.html'

> ⚠️ **Note:** This dataset is synthetic and was generated for modeling purposes.  
> In this case, the correlation between lead time and units sold is close to zero, indicating no visible relationship.

>  In real-world supply chains, **longer delivery times typically reduce sales**, especially in competitive industries.  
> The lack of impact here should be interpreted as a **simulation limitation**, not a business insight.


#  10. Category - Level Seasonality

### Monthly Seasonality by Product Category

To understand broader demand patterns, we analyze **average monthly sales** across all product categories.

This helps:
- Identify seasonal categories (e.g. beverages in summer)
- Benchmark SKU-level performance against category trends
- Support planning of promos or inventory

The chart below shows clear seasonal profiles for some categories.


In [25]:
# Prepare pivot
monthly_seasonality = (
    df.groupby(["month", "category"])["units_sold"]
    .mean()
    .unstack()
)

# Plot
plot_monthly_seasonality(monthly_seasonality)


'images/monthly_seasonality.html'


### Seasonality Patterns Across Product Categories

This line plot shows the **average monthly sales** (in units) by product category, helping us detect distinct seasonality behaviors across product types.
*  Juice and Yogurt show strong summer seasonality, peaking between June and August, likely driven by warmer weather and increased consumption of refreshing products.
*  ReadyMeals display the opposite trend, with higher demand in winter months (especially Q1), possibly linked to convenience-seeking behavior during colder seasons.
*  Milk has a more stable pattern with mild seasonality, suggesting regular, habitual consumption.
*  SnackBars show slightly higher activity around mid-year, but without a strong seasonal spike.

Implication:

These seasonal signals can help:

* fine-tune **promotion calendars**,
* optimize **production and inventory planning**, and
* personalize **marketing campaigns** based on seasonal product relevance.

 *Incorporating seasonality at the category level can improve forecast accuracy and increase business responsiveness to consumer behavior shifts.*



# 11. Feature Engineering 

###  for Forecasting

We engineer time-based features commonly used in time series forecasting, including:

- Lag features (1–2 weeks)
- Rolling mean and standard deviation (4 weeks)
- Weekly momentum (difference in recent lags)
- Target variable for next week (`target_next_week`)

This transformed dataset is saved as `weekly_df_final`.


In [26]:
# Posortuj dane po SKU / channel / region / week
weekly_df = weekly_df.sort_values(["sku", "channel", "region", "week"])

# Lag features: units_sold last week and two weeks ago
weekly_df["lag_1"] = weekly_df.groupby(["sku", "channel", "region"])["units_sold"].shift(1)
weekly_df["lag_2"] = weekly_df.groupby(["sku", "channel", "region"])["units_sold"].shift(2)

# Rolling windows (4 weeks): mean and std
weekly_df["rolling_mean_4"] = weekly_df.groupby(["sku", "channel", "region"])["units_sold"].transform(lambda x: x.shift(1).rolling(window=4).mean())
weekly_df["rolling_std_4"] = weekly_df.groupby(["sku", "channel", "region"])["units_sold"].transform(lambda x: x.shift(1).rolling(window=4).std())

# Momentum (WoW change)
weekly_df["momentum"] = weekly_df["lag_1"] - weekly_df["lag_2"]

# Target: units_sold in the following week
weekly_df["target_next_week"] = weekly_df.groupby(["sku", "channel", "region"])["units_sold"].shift(-1)

# Odfiltrowanie wierszy z brakami (na potrzeby modelowania)
weekly_df_final = weekly_df.dropna(
    subset=["lag_1", "lag_2", "rolling_mean_4", "rolling_std_4", "target_next_week"]
).copy()

# Podgląd finalnego datasetu
weekly_df_final.head()


,sku,week,channel,region,units_sold,stock_available,promotion_flag,price_unit,delivery_days,is_holiday_peak,week_number,month,year,is_holiday_week,is_summer,is_winter,sku_age,lifecycle_stage,lag_1,lag_2,rolling_mean_4,rolling_std_4,momentum,target_next_week
36,JU-021,2022-08-08,Discount,PL-Central,130,158.8,1,4.3820,3.000000,False,32,8,2022,0,1,0,4,Growth,122.0,99.0,132.25,27.121640,23.0,81.0
45,JU-021,2022-08-15,Discount,PL-Central,81,149.8,0,7.6140,4.000000,False,33,8,2022,0,1,0,5,Growth,130.0,122.0,125.00,20.704267,8.0,72.0
54,JU-021,2022-08-22,Discount,PL-Central,72,151.5,0,4.1525,2.500000,False,34,8,2022,0,1,0,6,Growth,81.0,130.0,108.00,22.286020,-49.0,125.0
63,JU-021,2022-08-29,Discount,PL-Central,125,134.0,1,3.9400,3.166667,False,35,8,2022,0,1,0,7,Growth,72.0,81.0,101.25,28.998563,-9.0,176.0
72,JU-021,2022-09-05,Discount,PL-Central,176,176.5,0,5.0300,4.000000,False,36,9,2022,0,0,0,8,Growth,125.0,72.0,102.00,29.743347,53.0,172.0


In [27]:
weekly_df_final.to_csv("weekly_df_final.csv", index=False)

#  12. Distributions & Outliers

### (SKU: MI-006)

Before modeling, it's helpful to visually inspect the **distributions** of engineered features such as:

- Target: future sales (`target_next_week`)
- Lags: past week sales (`lag_1`, `lag_2`)
- Rolling averages
- Momentum (% week-over-week change)

This allows us to:
- Detect skewed or non-normal distributions
- Spot outliers or rare values
- Identify features that may benefit from scaling or log transform


In [28]:
sku = "MI-006"
df_sku = weekly_df_final[weekly_df_final['sku'] == sku].copy()
selected_features = ['target_next_week', 'lag_1', 'lag_2', 'rolling_mean_4', 'momentum']

plot_feature_distributions_with_trend(df_sku, selected_features)


'images/feature_distributions_with_trend.html'


###  Feature Distribution Insight — Target & Predictors

To evaluate the quality and shape of key modeling features, we visualized the distribution of:

* `target_next_week` (future units sold — prediction target)
* `lag_1`, `lag_2` (previous week lags)
* `rolling_mean_4` (4-week moving average)
* `momentum` (week-over-week change)

---

###  Key Observations:

* All variables are **right-skewed**, reflecting the typical nature of **sales volume data** (many low/moderate values, few high spikes).
* The **target variable** and **lag features** show similar distribution shapes — confirming their predictive relevance.
* `rolling_mean_4` is smoother, as expected from temporal aggregation.
* `momentum` has a **more symmetric, bell-like shape**, centered near 0 — consistent with how change is calculated.

---

###  Why It Matters:

*  No obvious anomalies or multimodal patterns — features appear clean and reliable.
*  These plots support model selection and scaling decisions (e.g. tree-based models can handle skewed features without transformation).
*  You can now move confidently into **correlation analysis, feature importance, or model fitting**.

> The target and lag features show moderate skewness.  
> Momentum has some extreme outliers (±100%), suggesting either stockouts or promo shocks.  
> A log transformation or clipping may help stabilize modeling performance.


### Outliers (boxplot)

In [29]:
def detect_outliers_iqr(df, sku, feature, k=1.5):
    """
    Detects outliers for a given feature using the IQR rule.

    Parameters:
        df (pd.DataFrame): Weekly-level dataset
        sku (str): SKU to filter
        feature (str): Feature column to analyze
        k (float): Outlier multiplier (default: 1.5)

    Returns:
        DataFrame of outlier rows
    """
    df_sku = df[df["sku"] == sku].copy()
    Q1 = df_sku[feature].quantile(0.25)
    Q3 = df_sku[feature].quantile(0.75)
    IQR = Q3 - Q1
    lower = Q1 - k * IQR
    upper = Q3 + k * IQR

    outliers = df_sku[(df_sku[feature] < lower) | (df_sku[feature] > upper)]
    return outliers[[feature, "week"]].sort_values(by=feature)

# Example: detect outliers in momentum
detect_outliers_iqr(weekly_df_final, sku="MI-006", feature="momentum")


,momentum,week
2286,-161.0,2022-08-01
2185,-152.0,2022-05-16
2246,-151.0,2022-06-27
2234,-121.0,2022-06-20
2219,-110.0,2022-06-06
3235,-109.0,2024-08-05
2317,-102.0,2022-08-22
2291,-101.0,2022-08-01
2395,-96.0,2022-10-24
3271,-96.0,2024-09-02


### 🔍 Outlier Detection – Quantile Method (Target Variable)

To complement the visual analysis, we detect potential outliers in the target variable (`target_next_week`)  
using an extreme quantile range (0.5% on each side).

This identifies unusually low or high sales weeks that may reflect anomalies, campaign shocks, or supply issues.


In [30]:
# Define quantile cutoffs (lower and upper 0.5%)
q_low = weekly_df_final["target_next_week"].quantile(0.005)
q_high = weekly_df_final["target_next_week"].quantile(0.995)

# Identify potential outliers in the target variable
outliers = weekly_df_final[
    (weekly_df_final["target_next_week"] < q_low) |
    (weekly_df_final["target_next_week"] > q_high)
]

# Summary
print(f"Number of potential outliers: {len(outliers)}")

# Display top examples
outliers[["sku", "target_next_week", "rolling_mean_4", "momentum"]].head()


Number of potential outliers: 304


,sku,target_next_week,rolling_mean_4,momentum
1152,JU-021,5.0,77.25,54.0
1153,JU-021,23.0,60.25,-8.0
1154,JU-021,16.0,77.75,3.0
1155,JU-021,13.0,67.75,-2.0
1157,JU-021,26.0,55.00,-14.0


###  Observations & Recommendations

The histograms show reasonably well-distributed data, with a slight positive skew — which is typical in sales data.

Boxplots reveal the presence of outliers, but the number is not excessively high (e.g. 304 observations out of the full dataset).  


# 13. Summary & Business Recommendations

This analysis explored the weekly sales performance of **SKU `MI-006` (Milk)** in a simulated FMCG retail environment using daily-level data from 2022 to 2024.

The goal was to understand how various **operational and commercial factors** — such as promotions, stockouts, and seasonality — impact product demand over time.

---

### Key Insights:

- **Stockouts Matter:**  
  Even short periods of zero availability (e.g. 3–4 days) aligned with significant drops in sales.  
  One notable stockout streak occurred in May 2022 in the Retail channel.

- **Promotion Effectiveness Is Channel-Dependent:**  
  Promotions showed stronger uplift in the **Retail** channel compared to Discount or Online.  
  This highlights the importance of tailoring campaign strategy by channel.

- **Holiday Periods Amplify Promo Results:**  
  Promo events that aligned with peak shopping weeks (e.g. Christmas, Easter) led to higher incremental volume vs. promos run off-season.

- **Lifecycle Signals Exist:**  
  The SKU went through distinct **Growth → Maturity → Slight Decline** phases.  
  Promotional effectiveness was highest during the **Growth stage**, tapering in later stages.

- **Momentum & Trend Are Early Warning Signals:**  
  Week-over-week momentum detected inflection points before long-term declines set in.  
  This could help trigger proactive restocking or campaign planning.

- **No Clear Link Between Lead Time & Demand**  
  Due to synthetic nature of the data, longer delivery times did not show meaningful correlation with sales — though this would typically matter in real FMCG pipelines.

- **Outliers Exist but Are Manageable:**  
  Roughly 0.5% of observations fall outside expected target ranges. These can be:
  - Kept (robust models like LightGBM handle them well)
  - Capped or log-transformed if needed

---

###  Business Recommendations

✅ **Prioritize availability** during key calendar periods  
→ Even short stockouts reduce potential promo gains

✅ **Target promotions toward growth-stage SKUs**  
→ Avoid overspending on promos for saturated or declining products

✅ **Use channel-specific planning**  
→ Retail shows highest response to promotions for this SKU

✅ **Track weekly momentum early**  
→ It provides leading signals for forecasting or campaign design

✅ **Reassess lead-time impact in real-world data**  
→ Consider exploring this further with operational sources

---

> ⚠️ _Reminder: This dataset is synthetic and was created for portfolio demonstration purposes.  
> Patterns like flat lead-time impact or limited holiday interaction may differ in real FMCG data._  
